### ISSUE TITLE
Support using plugins to extend scheduler 

### ISSUE THREAD

**Issue by cwen0** [chaos-mesh/chaos-mesh#193]
## Description

Sometimes we need to customize our selector. For example, when our cluster has multiple pds, we just want to inject the failure to the leader.

We can provide a plugin to extend the selector, and users can implement their own selector through code. For example, we implement a file called pd-selector.go, we can set it in the selector

```
selectors:
    plugin: "pd-selector"
```  

## Recommended Skills

* Golang
* [CRD](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)

**Comment by zhouqiang-cl** [comment]
Make sense!!!

**Comment by AngleNet** [comment]
Maybe we should support some builtin selector plugins, users can choose one of them. I think we could support to load a third-party selector plugin which comply to our selector plugin specification as a cutomized selector for a chaos. @cwen0

**Comment by cwen0** [comment]
@AngleNet Yes, a third-party selector plugin is ok and I think this plugin not only for the selector in the future. Are you interested in this issue?

**Comment by AngleNet** [comment]
Yes. I’m working on another issue for now. When it’s finished, you can assign it to me

**Comment by vincent178** [comment]
Hi @cwen0 @AngleNet can I work on this? Thanks.

**Comment by zhouqiang-cl** [comment]
Hi, @vincent178 How about the progress now

**Comment by vincent178** [comment]
Sorry didn't start work on this, feel free to pick it up. Otherwise I can speed some time this week.

**Comment by zhouqiang-cl** [comment]
@vincent178 take your time, waiting for your PR😃

**Comment by vincent178** [comment]
Hi @cwen0 @zhouqiang-cl , after some investigation, I have two ideas of this task:

## Use go plugin 
Go has a way to execute code dynamically with https://golang.org/pkg/plugin/.
A plugin file would looks like:
```go
package main // this has to be a main package
func Select(context.Context, client.Client) ([]v1.Pod, error) {
}
```
My rough idea would be add a new folder under `pkg` called `plugins`, add builtin plugins to it, compile to *.o file and execute in `pkg/utils/selector.go` when end user specify this plugin in the yaml file. 
We can also add ability to support different plugin load path, so end user can write their own plugin, compile to *.o file and add it to docker image and passing the plugin path when start the docker container.

## Use in-code pluginfactory
We can build a `PluginFactory` which holds all the factory name and function mapping, like 
```go
type PluginFactory struct {
  plugins map[string]func(context.Context, client.Client) ([]v1.Pod, error)
}
```
and end user write a plugin, register the plugin through code, and specify this plugin in yaml file.


I prefer using the first options, because the end-user have the ability to not write the code inside chaos-mesh codebase, but only copy the *.o file into the existing chaos-mesh image, while the second option, the end user has to write the code, build the image from scratch. What do you guys think?

**Comment by YangKeao** [comment]
@vincent178 Thanks for your idea :smile: 

I don't think go plugin is a good idea. Here I will list several reasons:

1. The plugin should have the same compiler version with the main binary. It means that if our `controller-manager` is compiled with go 1.14.6, it won't work to compile the plugin with go 1.14.5.

2. Any packages used by both the main binary and plugin must have their version match exactly. 

Go plugins are designed to be used in an in-tree scenario, where all codes share the same building environment and the same `go.mod` file. In real `plugins` situation, it will bring you a lot of headaches.

A normal plugin design can be starting a new process and communicate with it (by UNIX sockets or stdin/stdout). For example, to launch plugin `pd-selector`, we can execute a program named `pd-selector`, read/write the stdin/stdout according to some protocol (`jsonrpc` is a good choice), and get the result.

The plugin can be one time (will be executed and closed every time) or be a daemon, both of them have pros and cons.

**Comment by vincent178** [comment]
@YangKeao thanks for your reply, the Cons of using go plugins make sense to.

I wonder if we use the binary way, how should this binary deployed into chaos-mesh or k8s clusters?

**Comment by YangKeao** [comment]
> @YangKeao thanks for your reply, the Cons of using go plugins make sense to.
> 
> I wonder if we use the binary way, how should this binary deploy into chaos-mesh or k8s clusters?

In the same way as how you copy the *.so into the container :sweat_smile:.

Building a new image is one of the approaches, or maybe we can prepare a repo to collect and compress some common "official" plugins.

**Comment by vincent178** [comment]
I see, I was thinking using some k8s way to do it, I think just copy into container is better. Let me think about it, thanks again.

**Comment by YangKeao** [comment]
@vincent178 Great suggestion! In k8s way, maybe we could provide a configuration file (like a Service in k8s) to setup plugin.

1. The user setups plugin pods and services.

2. Register these pods/services with a "ChaosMeshSelectorPlugin" resource.

3. When the controller needs to call the plugin, communicate with it through Grpc or other Rpc over the network.

This way is much more flexible and elegant. I think the cost of the network is definitely acceptable, but it's more unstable than local communication (through unix socket or stdin/stdout).

**Comment by STRRL** [comment]
I still have a little confusing about why we need a plugin other than label-selector. 

For example, a plugin could find out which one is the leader of PD cluster then hint scheduler injecting failures; Or just tag label like `role: leader` to the leader pod, then using a Chaos CRD with label-selector `role: leader`.

**Comment by YangKeao** [comment]
> I still have a little confusing about why we need a plugin other than label-selector.
> 
> For example, a plugin could find out which one is the leader of PD cluster then hint scheduler injecting failures; Or just tag label like `role: leader` to the leader pod, then using a Chaos CRD with label-selector `role: leader`.

(I agree with you :sweat_smile: .) @cwen0

**Comment by cwen0** [comment]
> I still have a little confusing about why we need a plugin other than label-selector.
> 
> For example, a plugin could find out which one is the leader of PD cluster then hint scheduler injecting failures; Or just tag label like `role: leader` to the leader pod, then using a Chaos CRD with label-selector `role: leader`.

Compared with writing a separate program to set the label of application, the plug-in has the following advantages:

1. Don't need to deploy and maintain the separate processes, `plugin` is more user-friendly
2. It is easier to develop a chaos mesh plugin compare to a separate process. Because we can provide some functions to help uses to develop plugin, such as `GetPods(), GetNamespace...` 
3. The plugin can be reused and avoid duplication of development work

**Comment by STRRL** [comment]
Thanks for your explain. It does be helpful to developing with some plugins mechanism.

It seems need a lot of work on this: plugin API, build plugin and how to execute my customized plugin on an already existed chao-mesh.

Expecting for a well-designed plugin framework.😁

**Comment by yeya24** [comment]
Closed by accident. Still valid and help wanted.
